### DISCUSSION TITLE
Inaccurate SID Emulation

### DISCUSSION THREAD

**Discussion by Brog33** [Alexco500/sidplay5#15] [General]
I have noticed a playback issue in sidplay5 which has been fixed in the later versions of VICE (Perhaps VICE is using a newer version of reSID?) I can provide an example if you like.

**Comment by Alexco500** [comment]
Yes, please do. Sidplay5 still uses the old libsidplay2 with an older reSid. I am still trying to figure out if it would be feasible to switch completely to the new libsidplayfp lib. Most likely, I would lose some functions from the info window...

**Reply by Brog33** [reply]
Ok, here's an example. Bit of a weird tune this one, but especially the bass (appears after 12 seconds) sounds bad in sidplay. In VICE and on real C64 it sounds fine.

[Drive_on_Music.sid.zip](https://github.com/Alexco500/sidplay5/files/11332459/Drive_on_Music.sid.zip)

**Comment by Alexco500** [comment]
Hi @Brog33, yes, that is a weird tune. It took me a while to hear the difference. To fix that, I would need to switch to newer version of reSID (official repo), reSID (VICE repo) or use reSIDfp. Did also try sidplayfp?
Regardless of what I choose, sidplay will lose the filter settings in the info window, since reSID significantly changed in that part.

**Reply by Brog33** [reply]
i have not tried sidplayfp, i could not see a macos version of it. maybe you know?

what sid filter settings are you refering to? this one?
![image](https://user-images.githubusercontent.com/1303537/235320204-465a0c4c-aaa1-41f9-98d7-cf8af5bd5d22.png)

**Comment by Alexco500** [comment]
No, I compiled sidplayfp by myself.

I meant this one (select a song and press cmd+i):
![grafik](https://user-images.githubusercontent.com/20357513/235321351-10ba55bf-24b0-4fa7-8a58-e24ba332e0d2.png)

**Reply by Brog33** [reply]
haha wait.. wow, how? i have never discovered this window.. this app is so puzzling :D i think the UI needs a button/icon to open this info window :) 

i gotta say this is **very cool**! although personally i would stick with the default SID filter settings.

**Reply by Alexco500** [reply]
Yes, it is really a nice feature. But these filter settings are no longer supported by newer reSID versions, at least I can't see how I could implement that. On the other hand, I never play around with that, I just listen to the music...

**Comment by Alexco500** [comment]
Can you check latest Beta release please?

https://github.com/Alexco500/sidplay5/releases/tag/v5.1.0_BETA1

**Reply by Brog33** [reply]
Nice, now the SID toggler is visible :) 
Unfortunately 'Drive on Music' still sounds buggy in this version. 

Btw: The about window still shows 5.0.1 (44) while the window title shows 5.1 BETA

**Comment by Alexco500** [comment]
Really?!?!? The sound is created by the reSID component, so I thought by replacing that with the latest and greatest version, the sound issue is solved.
I guess then I need to switch to libsidplayfp completely. That may take a while..

**Reply by Brog33** [reply]
i hear a difference between the two versions, but still that one SID-tune doesn't sound correct 🤷‍♂️

it would be very interesting to listen to that SID-tune with sidplayfp. since you have sidplayfp compiled, perhaps you could listen to it? my reference is using VICE 3.7 to play that SID-tune as well as a real C64.

**Reply by Alexco500** [reply]
Puh. Frankly, I can't hear a difference. What I did: I listen to the tune with sidplay, then with sidplayfp, and then I load the prg (the exported via sidplay as C64 prg) into VICE 3.7 and Denise. I can't listen to it on a real C64, mine is broken and still under repair. I could use my MEGA65, but that would only test the SID FPGA implementation.
Furthermore, I even listened to it with my SIDBlaster USB, but sorry to say, I really can't hear the difference :-(.

Is there any further hint you can give to me, like a special time range? Or maybe a different song? Then I could try to export that and compare more easliy.

**Reply by Brog33** [reply]
I have cut out a section of the tune that should (hopefully) be able to identify the differences. These are all using 6581
There are two things to pay attention to here. 
1. The bass - it sounds 'not clean' in sidplay.
2. At 22 seconds there's a glide which goes out-of-whack in sidplay.

I can try to gather some more examples if you need.

[DOM.zip](https://github.com/Alexco500/sidplay5/files/11404682/DOM.zip)

**Comment by Alexco500** [comment]
No, thanks, that makes it pretty clear, even for me :-)
In the meantime, I switched to libsidplayfp, using reSIDfp [(link)](https://github.com/Alexco500/sidplay5/releases/tag/v5.1.0_BETA2).

A lot of things won't work, e.g. time, info view, ... and some may even crash.
The playback still sounds not completely okay, but I guess you will find out :-D

**Reply by Brog33** [reply]
I don't know why Drive on Music still sounds weird, but I really appreciate your work on this.

**Comment by Alexco500** [comment]
@Brog33 
Did you check latest Beta (https://github.com/Alexco500/sidplay5/releases/tag/v5.1.0_BETA4)?
It is using reSID with VICE default settings (filter, bias, ...), not reSIDfp as previous beta.

**Comment by Brog33** [comment]
Testing v5.1.0_61 now, it has some problems with playback (same with v5.1.0_BETA4 posted above).  
Here is a good example, you can hear after 6 seconds it has problems (made for 8580).
[All_Around_the_World.sid.zip](https://github.com/Alexco500/sidplay5/files/12643344/All_Around_the_World.sid.zip)

In your 5.1 reSID beta it sounds fine.

**Reply by Alexco500** [reply]
Hm, strange. I use the same default settings for the reSID engine as VICE. Is there any BETA version which does not show playback errors? I mean, you tested already a lot of SIDs, but it seems that there is no version which plays them all correctly. Just for reference:

- Beta 1: updated reSID engine, old sidplay library
- Beta 2: updated reSID engine, new sidplayfp library
- Beta 3: switch to reSIDfp engine, same sidplayfp as Beta 2
- Beta 4: switch to reSID engine (same as Beta 1 and 2) with VICE settings, sidplayfp library

If the SID plays fine with Beta 1, can you test that with Beta 2? Maybe there is a sidplayfp lib issue?

**Comment by Brog33** [comment]
i have currently these, which i have tested on:
- SIDPLAY 5_1 BETA (libsidpf_reSIDfp)
- SIDPLAY 5_1 BETA (reSID)
- SIDPLAY 5_1_0 BETA 4 (sidplayfp)
- SIDPLAY (release version)

( SIDPLAY 5_1 BETA (reSID) is the only of them that plays it correctly. )

I am not sure which of these corresponds to your list, do you have link to all the betas so I can referer to the correct beta number?

**Reply by Alexco500** [reply]
This should be Beta 1, then. This means that libsidplayfp has some issues with that.
For the list of betas, you can use this one: https://github.com/Alexco500/sidplay5/tags 
Simply click on one of the versions for the description.

**Comment by Brog33** [comment]
nice, now i've downloaded them again and set the correct name so i can test properly.

so yeah, for "all around the world": beta 1 sounds correct. beta2 & 3 doesn't sound quite right. beta 4 sounds bad (high pitched sounds after 6 seconds).

"drive on music" sounds incorrect in all 4 betas, though. this seems to be a particular tricky one. VICE is able to play it correctly, but no sid player that I've tried could do it.

**Reply by Alexco500** [reply]
Can you please try latest release (v5.1.0_66)? It contains latest libsidplayfp and reSID.

**Reply by Brog33** [reply]
'All around the world' sounds good with 5.1.0_66 👍 however when I exported it as .aiff for external playback it sounded very glitchy. 

'Drive on Music' still sounds messed up.

**Comment by Brog33** [comment]
in case you wonder how the exported audio sounds:
[sidplay.aiff.zip](https://github.com/Alexco500/sidplay5/files/12726814/sidplay.aiff.zip)

**Reply by Alexco500** [reply]
That sounds awful… but I guess I found the issue. 
As for the ‘drive on music’ tune, I guess I open an issue for libsidplayfp. I really don’t know how to fix that.

**Reply by Alexco500** [reply]
Please test the new release here
https://github.com/Alexco500/sidplay5/releases/tag/v5.1.0_67

**Reply by Brog33** [reply]
Same glitchy sound in the AIFF export in the new version as well.

**Reply by Alexco500** [reply]
Really? I tried now more than 10 tunes, all okay. What I do: mark a tune either in a playlist or collection, context menu "Export as aiff". Work so far without issues.

**Reply by Brog33** [reply]
Did you try to export the tune All around the world specifically?

**Reply by Alexco500** [reply]
Yes, here is the result (50s only).

[Jens-Christian Huus - All Around the World (song 1).aiff.zip](https://github.com/Alexco500/sidplay5/files/12841451/Jens-Christian.Huus.-.All.Around.the.World.song.1.aiff.zip)

**Comment by Brog33** [comment]
To reproduce:
1. Make sure all instances of sidplay is fully terminated.
2. Open the app.
3. Do not start playing any SID - Just open the AIFF export window for all around the world. 
4. Do not alter any settings here, keep the 03:00 time limit (if you set it to 00:50 the bug will not occur).
5. Export

Hopefully you will get the glitch now :) (I could reproduce this several times)

**Reply by Alexco500** [reply]
Ha! You're right, thank you very much!
If you don't use the menu, but the context menu (right click) on the song, it should work.
Time for a fix, I guess...

**Reply by Alexco500** [reply]
Please check latest release, v5.1.0_73.

**Reply by Alexco500** [reply]
Okay, I see that it is still not fixed completely. But the issue is not entirely connected to export only. If you do the export like described above, it works (at least here).
But if you then try to play the song in the player, it will be distorted again. Playing a different song works. So more testing needed :-(

**Reply by Alexco500** [reply]
Please check release v5.1.0_74. I was not able to reproduce the glitches again...

**Reply by Brog33** [reply]
now it works 👍
